### PR TITLE
[Anitspam Idea] Sidekiq IPSpamlistURLScheduler

### PR DESCRIPTION
**Pull request by shleeable**
Mastodevs, even if this doesn't get committed.. could you please check the code and feel free to recommend changes?

I'll likely push this to my fork if the code is at least correct and safe.

fyi I have no idea how to add tests for this kind of thing.

### CODE REVIEW

**Comment by mjankowski** [comment]
Vaguely related - https://github.com/mastodon/mastodon/pull/32981

This has been open ~10 months and no feedback yet. Should we leave open, or close?

Separately...

> could you please check the code and feel free to recommend changes?

Can you rebase this against main and I'll give you a review? Not sure if we can get it mergeable or not, but I'll give feedback.

> I have no idea how to add tests for this kind of thing.

High level you'd want to stub out the http responses to give you some IPs back in whatever format is relevant, and then assert that running the job leads to the relevant updates on the ip block list side of things...

File: app/workers/scheduler/ip_spamlist_urls_scheduler.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+class Scheduler::IpSpamlistUrlsScheduler
+  include Sidekiq::Worker
+
+  sidekiq_options retry: 0
+
+  def perform
+    grab_ip_blocks_addresses! if ENV['SCHEDULED_IPBLOCK_URLS'].present?
```
Comment (mjankowski) [inline app/workers/scheduler/ip_spamlist_urls_scheduler.rb:9]: Minor style thing, but -- I'd probably just have either one top level thing (`process_ip_urls!` or something) or two (one for limit, one for block), and then have those methods wrap up the "if env var present" loop)

File: config/sidekiq.yml [hunk 2]
```diff
@@ -67,3 +67,7 @@
       interval: 1 hour
       class: Scheduler::AutoCloseRegistrationsScheduler
       queue: scheduler
+    ip_spamlist_url_scheduler:
```
Comment (mjankowski) [inline config/sidekiq.yml:70]: Do we need to contemplate the scenario where a person wants to NOT BLOCK an IP from these lists?

File: app/workers/scheduler/ip_spamlist_urls_scheduler.rb (hunk 1 above)
Comment (shleeable) [reply app/workers/scheduler/ip_spamlist_urls_scheduler.rb:9]: I'm not a programmer but my intention is purely readability.

The first function pulls the lists from the array of 1 or more URLS and appends those ips to be processed.
the second function processes the ips.

I don't understand ruby enough to fix this... I hacked my way here and I don't remember how

File: app/workers/scheduler/ip_spamlist_urls_scheduler.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        @limitips.insert = res.body
+      end
+    end
+  end
+
+  def add_ip_blocks_addresses!
+    @blockips.each do |ip|
+      ip_block = IpBlock.find_by(ip: ip)
+
+      if ip_block.present?
```
[the remark this answers is not in the document]
Comment (shleeable) [reply app/workers/scheduler/ip_spamlist_urls_scheduler.rb:37]: correct.

**Comment by shleeable** [comment]
I don't understand ruby enough to fix this... I hacked my way here and I don't remember how.
This is no longer my code because it was done by me in the past. I'd need somebody to take it over.

**Comment by mjankowski** [comment]
> I don't understand ruby enough to fix this... I hacked my way here and I don't remember how.

🫡 

> This is no longer my code because it was done by me in the past. I'd need somebody to take it over.

Sounds like the premise of some sort of alternate timelines programming-themed science fiction, where a developer is battling another developer who keeps changing their code ... only to discover that THE CALL IS COMING FROM INSIDE THE HOUSE -- that's right, it was them all along! Them-of-the-past has been breaking the build!

All that said, I'm going to close this on the assumption that zero response in ~10 months means no one wants to shepherd this into the future.

I'm marginally hopefully that some recent antispam work (see commits last few days) might evolve to make something like this more do-able.

**Pull request closed without merging**